libyul/optimiser/KnowledgeBase.h

```diff
@@ -37,15 +38,16 @@ struct AssignedValue;
 
 /**
  * Class that can answer questions about values of variables and their relations.
- *
- * The reference to the map of values provided at construction is assumed to be updating.
  */
 class KnowledgeBase
 {
 public:
-	KnowledgeBase(Dialect const& _dialect, std::map<YulString, AssignedValue> const& _variableValues):
+	KnowledgeBase(
+		Dialect const& _dialect,
+		std::function<std::map<YulString, AssignedValue> const&()> _variableValues
```

Comment:
I wonder if it wouldn't actually be nicer to make this a
``std::function<AssignedValue const*(YulString)>``

But this would require further changes down to ``SimplificationRules`` and ``Pattern``, so maybe not worth it at least right away...

Reply:
CommonSubexpressionEliminator.cpp:109 does the traversal

Reply:
Ah wait, but that's a subclass